### ISSUE TITLE
[20.09] File download seems non-functional

### ISSUE THREAD

**Issue by selten** [galaxyproject/galaxy#10393]
I deployed Galaxy 20.09 to my playground environment today, however downloading of files from Galaxy doesn't seem to work.

Hereby the logs that I got:
```
IP - - [09/Oct/2020:16:21:59 +0200] "GET /datasets/DATASETSTRING/display?to_ext=txt HTTP/1.1" 500 - "https://GLXURL/"
Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/error.py", line 153, in __call__
    app_iter = self.application(environ, sr_checker)
  File "GLXPATH/.venv/lib/python3.7/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "GLXPATH/.venv/lib/python3.7/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 141, in __call__
    return self.handle_request(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 220, in handle_request
    body = method(trans, **kwargs)
  File "lib/galaxy/webapps/galaxy/controllers/dataset.py", line 186, in display
    return data.datatype.display_data(trans, data, preview, filename, to_ext, offset=offset, ck_size=ck_size, **kwd)
  File "lib/galaxy/datatypes/data.py", line 440, in display_data
    trans.response.headers['Content-Length'] = int(os.stat(data.file_name).st_size)
FileNotFoundError: [Errno 2] No such file or directory: ''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/batch.py", line 80, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/middleware/request_id.py", line 15, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/xforwardedhost.py", line 23, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/translogger.py", line 70, in __call__
    return self.application(environ, replacement_start_response)
  File "lib/galaxy/web/framework/middleware/error.py", line 163, in __call__
    exc_info)
  File "lib/galaxy/web/framework/middleware/translogger.py", line 69, in replacement_start_response
    return start_response(status, headers, exc_info)
SystemError: <built-in function uwsgi_spit> returned a result with an error set

```

**Comment by mvdbeek** [comment]
That looks like the file has been purged. See if this file is still on disk ?

**Comment by selten** [comment]
The file is no longer on disk, but wasn't purged by me either. 

The file is at path:
```
GLXPATH/database/files/001/339/dataset_XXXXX.dat
```

However, even `GLXPATH/database/files/001/339/` no longer exists.

**Comment by selten** [comment]
So it was also already deleted before I updated today, so maybe it's one of the cleaning scripts that deleted it.

**Comment by selten** [comment]
Okay, seems like this was caused by a change in the default path of files from database/files to objects.

**Comment by scholtalbers** [comment]
I have a similar issue after updating from 20.01 to **20.05**. The files are however still available on disk e.g. `/g/galaxy-production/database/files/000/375/dataset_375280.dat` but the download or preview fails with the above mentioned error (see below as well). 

The history or dataset is not deleted - at least not marked as such as far as I can tell. PNG, bigwig, txt files cannot be previewed or downloaded, but BAM files in the same history can. Also newly uploaded or generated PNGs do seem to work.

After reading the above comments I updated my config to explicitly set `data_dir` to `database/`, but that didn't change anything (I guess that would only make a difference for 20.09).

```
404 Not Found
The resource could not be found.

File Not Found ().
```
and

```
 - [28/Oct/2020:09:11:35 +0200] "GET /datasets/cc15cf49bb47771b/display?to_ext=png HTTP/1.1" 500 - "https://galaxy.embl.de/datasets/edit" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.80 Safari/537.36"
Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/error.py", line 154, in __call__
    app_iter = self.application(environ, sr_checker)
  File "/g/galaxy-production/.condavenv/lib/python3.6/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "/g/galaxy-production/.condavenv/lib/python3.6/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 145, in __call__
    return self.handle_request(environ, start_response)
  File "lib/galaxy/web/framework/base.py", line 224, in handle_request
    body = method(trans, **kwargs)
  File "lib/galaxy/webapps/galaxy/controllers/dataset.py", line 187, in display
    return data.datatype.display_data(trans, data, preview, filename, to_ext, offset=offset, ck_size=ck_size, **kwd)
  File "lib/galaxy/datatypes/data.py", line 451, in display_data
    trans.response.headers['Content-Length'] = int(os.stat(data.file_name).st_size)
FileNotFoundError: [Errno 2] No such file or directory: ''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "lib/galaxy/web/framework/middleware/batch.py", line 80, in __call__
    return self.application(environ, start_response)
  File "lib/galaxy/web/framework/middleware/request_id.py", line 15, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/xforwardedhost.py", line 23, in __call__
    return self.app(environ, start_response)
  File "lib/galaxy/web/framework/middleware/translogger.py", line 71, in __call__
    return self.application(environ, replacement_start_response)
  File "lib/galaxy/web/framework/middleware/error.py", line 164, in __call__
    exc_info)
  File "lib/galaxy/web/framework/middleware/translogger.py", line 70, in replacement_start_response
    return start_response(status, headers, exc_info)
SystemError: <built-in function uwsgi_spit> returned a result with an error set
```

**Comment by cgirardot** [comment]
Hi @mvdbeek 
sorry to mention you directly but I'd like to make sure you saw @scholtalbers comment (not sure how this works when the issue is already closed). Indeed, this problem is general as we got complains from many users that they can't see/download their data. Newly computed data is accessible though. We would really appreciate if you could re-open the issue. Again sorry to insist.

**Comment by mvdbeek** [comment]
If the file `/g/galaxy-production/database/files/000/375/dataset_375280.dat` is available, and the path is constructed properly, it will be downloadable.

> After reading the above comments I updated my config to explicitly set `data_dir` to `database/`,

https://docs.galaxyproject.org/en/master/admin/options.html#data-dir that's fine, but shouldn't matter.
You might want to try setting `object_store_store_by` to `id`. This should normally be set to `id` by Galaxy if an existing instance updates to 20.05.
The logic there is https://github.com/galaxyproject/galaxy/blob/ac0b41abf0fc6e04ab3e9b23c8f98e297be4a391/lib/galaxy/config/__init__.py#L708

**Comment by cgirardot** [comment]
thx @mvdbeek for the quick feedback. The file is definitely here, with correct permissions: 
```
$ ls -l /g/galaxy-production/database/files/000/375/dataset_375280.dat 
-rw-r--r-- 1 galaxy gbcs 6288369 Oct 13 18:06 /g/galaxy-production/database/files/000/375/dataset_375280.dat
``` 

it seems we don t have any `object_store_store_by` set in any of our config files. In which config file should I add this ? Would `galaxy.yml` be the one ?
object_store_store_by: id

**Comment by mvdbeek** [comment]
Yes, unless you have multiple object stores where you mix `id` and `uuid`, in which case you can set this using the `store_by` parameter on the <object_store> element.

**Comment by cgirardot** [comment]
we don t have any object stores defined. I added `object_store_store_by: id` under the `galaxy` section of `galaxy.yml` (restarted galaxy) but the problem is still here

**Comment by mvdbeek** [comment]
Are you sure you are looking at the `HistoryDatasetAssociation` (what you see in the history) that is linked to the `Dataset` 37528 ? These are not the same thing, `375280` is the dataset id for the file you're showing on disk. If you have the id of a `HistoryDatasetAssociation` you can get the dataset id in the database with `SELECT dataset_id FROM history_dataset_association WHERE id = <your numeric HDA id>;`

**Comment by cgirardot** [comment]
I am pretty sure, here is a screenshot showing the command line that was exec for the dataset (png) that does not display
![info](https://user-images.githubusercontent.com/10990795/97432481-4c035e80-191c-11eb-9946-9e2680058d2d.png)

**Comment by cgirardot** [comment]
and if I open the file, I see the expected heatmalp

![Screenshot 2020-10-28 at 12 54 51](https://user-images.githubusercontent.com/10990795/97432774-c9c76a00-191c-11eb-9d71-623ed1244794.png)

**Comment by mvdbeek** [comment]
Do you have more than one object store setup ? Is the path where you drew a red rectangle not `fungen` ?

**Comment by cgirardot** [comment]
we have only one. All files are on the same NFS share. I was only trying to hide some info on the exact path but I miserably failed 🙈

**Comment by mvdbeek** [comment]
Can you import the history/dataset into an admin account and see what is written in the UI under `Full Path:` and compare that with a dataset that works?

**Comment by cgirardot** [comment]
ah . Full Path is empty !

**Comment by cgirardot** [comment]
and it is correctly set for a dataset that I generated yesterday

**Comment by mvdbeek** [comment]
What is the path to that dataset ?

**Comment by cgirardot** [comment]
`/g/funcgen/galaxy-production/files/000/380/dataset_380648.dat`

**Comment by mvdbeek** [comment]
OK, so this dataset uses `/g/funcgen/galaxy-production/files` while your old datasets use `/g/funcgen/galaxy-production/database/files`. I think at this point you could either merge the two locations if there is no conflict, or setup 2 object stores with different paths.

**Comment by cgirardot** [comment]
outch I see. Not sure what this hapenned, I will look in this object store story. I guess I should define a `object_store_conf.xml` based on `object_store_conf.xml.sample` ?

**Comment by cgirardot** [comment]
sorry, jelle is off today and he usually deals with the configs. Your "merge" option should also be good as we have very few new datasets under this `/g/funcgen/galaxy-production/files` path. 
In `galaxy.conf` I see : 
   `file_path: files` 
I assume this should then be : 
   `file_path: database/files`

Just to make sure I get this right, I should :
- switch off galaxy
- change this `file_path` to `database/files` 
- move the few files from `/g/funcgen/galaxy-production/files` to `/g/funcgen/galaxy-production/database/files`
- then restart 

correct ?

**Comment by mvdbeek** [comment]
yes, I think that looks right. You can also first change the setting and restart, then move the files into place and check that after the move they are indeed available for download and display.

**Comment by scholtalbers** [comment]
To comment on this, I updated the instance without updating `galaxy.yml`. 
`data_dir` and `file_path` were unset
`data_dir` should default to `database` and `file_path` should be relative to this - at least from the comment `data_dir` comment. In my memory the default relative `file_path` would always resolve to `database/files`.
```
# The directory that will be prepended to relative paths in options
  # specifying Galaxy data/cache directories and files (such as the
  # default SQLite database, **file_path**, etc.). Defaults to `database/`
  # if running Galaxy from source or `<config_dir>/data` otherwise.
  #data_dir: null
```

**Comment by scholtalbers** [comment]
FYI the `file_path: database/files` fixes the issue. I'm currently copying/moving the files from `$galaxyhome/files` to the correct `$galaxyhome/database/files`.

**Comment by cgirardot** [comment]
thx a bunch @mvdbeek

**Comment by ic4f** [comment]
> FYI the `file_path: database/files` fixes the issue. 
The `database/` prefix should happen automatically for `files`. This doesn't seem right. I'm reopening the issue to check that this is resolved correctly, and whether we haven't missed anything in the recent updates to path resolution.

**Comment by ic4f** [comment]
Yep, I can reproduce the download error. Working on a fix.

**Comment by ic4f** [comment]
This issue has been fixed in #10617. 

Please let us know if you experience any problems with the `file_path` and/or `object_store_store_by` settings. We've tried to cover all the main cases; however, when setting the defaults for these options Galaxy has to make a guess: (1) when `object_store_store_by` is not set by the admin, and (2) when a `<data_dir>/files` directory already exists - so it is possible that it may guess incorrectly. Also, please see the updated documentation for both options:
https://github.com/galaxyproject/galaxy/blob/release_20.09/lib/galaxy/config/sample/galaxy.yml.sample#L648
https://github.com/galaxyproject/galaxy/blob/release_20.09/lib/galaxy/config/sample/galaxy.yml.sample#L215
